# Introduction To Fn

## 어떤 밋업 갔다가 알게 되 opensource라서 한번 돌려 보고 싶어서 여기에 쓰는 것임

## 1. install jupyter kernal for golang

### Jupyter에 golang kernel 설치 먼저 함.

origin project url:  https://github.com/fnproject/tutorials/blob/master/Introduction/README.md

```bash
go get -u github.com/gopherdata/gophernotes

mkdir -p ~/.local/share/jupyter/kernels/gophernotes

cp $GOPATH/src/github.com/gopherdata/gophernotes/kernel/* ~/.local/share/jupyter/kernels/gophernotes  
```

**참조** GOPATH가 글로벌로 설정되어있지 않다면 GOPATH 위치를 알아야 하는데 다음과 같은 명령어를 치면 된다.
```golang
go env |grep GOPATH
```
jupyter 재시작해서 커널이 제대로 읽어오는 지 확인 필요!!
---

만약 go get -u github.com./gopherdata/gophernotes를 할 때 
'**libzmq was installed under /usr/local/lib,
/usr/local/lib/pkgconfig/libzmq.pc exists.
PKG_CONFIG_PATH sets to /usr/local/lib/pkgconfig**' 
처럼 나올 때, 

https://stackoverflow.com/questions/24250838/problems-installing-gozmq-with-go-get 참고함


아래 사이트 가서 libzmq를 받아 컴파일 했고, compile에 필요한 **build-essential, libtool, autoconf, automake, and uuid-dev**을 설치함
http://zeromq.org/docs:source-git
```bash
sudo apt install automake
sudo apt install autoconf
sudo apt install libtool
```


### "Kernel error" in a running notebook

커널 에러가 나서 문제가 먼가 했는데, gophernotes가 심볼릭 링크를 못 찾아서 그런 것!
그리고 이런 문제가 많았던지 go kernel 설치하는 github에 README.md 마지막 줄에 아래와 같이 있음.

```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/notebook/base/handlers.py", line 458, in wrapper
    result = yield gen.maybe_future(method(self, *args, **kwargs))
  File "/usr/local/lib/python2.7/site-packages/tornado/gen.py", line 1008, in run
    value = future.result()
  ...
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1335, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```

Stop jupyter, if it's already running.

Add a symlink to `/go/bin/gophernotes` from your path to the gophernotes executable. If you followed the instructions above, this will be:

```
sudo ln -s $HOME/go/bin/gophernotes /go/bin/gophernotes
```

Restart jupyter, and you should now be up and running.

## 2. Docker Install

In [1]:
!curl -fsSL https://get.docker.com/ | sudo sh

# Executing docker install script, commit: 1d31602
+ sh -c apt-get update -qq >/dev/null
+ sh -c apt-get install -y -qq apt-transport-https ca-certificates curl >/dev/null
+ sh -c curl -fsSL "https://download.docker.com/linux/ubuntu/gpg" | apt-key add -qq - >/dev/null
+ sh -c echo "deb [arch=amd64] https://download.docker.com/linux/ubuntu xenial edge" > /etc/apt/sources.list.d/docker.list
+ [ ubuntu = debian ]
+ sh -c apt-get update -qq >/dev/null
+ sh -c apt-get install -y -qq --no-install-recommends docker-ce >/dev/null
+ sh -c docker version
Client:
 Version:	17.12.0-ce
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	c97c6d6
 Built:	Wed Dec 27 20:11:19 2017
 OS/Arch:	linux/amd64

Server:
 Engine:
  Version:	17.12.0-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	c97c6d6
  Built:	Wed Dec 27 20:09:53 2017
  OS/Arch:	linux/amd64
  Experimental:	false
If you would like to use Docker as a non-root user, you should now consider
adding your user to th

## 3. Fn Download and Install

In [4]:
!curl -LSs https://raw.githubusercontent.com/fnproject/cli/master/install | sh

fn version 0.4.34

        ______
       / ____/___
      / /_  / __ \
     / __/ / / / /
    /_/   /_/ /_/`



## 4. Fn start

```bash
$ sudo fn start
```

![FN 실행](fn_start.png)

---
## 5. Fn run
### 간단히 go를 통해 예제 돌려 보자!

* go 파일 만들기
  ```bash
    mkdir hello
    cd hello
  ```
* 아래 텍스트를 func.go 생성 후 삽입.
  ```go
    package main
    import {
      "fmt""
    }
    func main(){
      fmt.Println("Hello World!")
    }
  ```
* yaml 생성
  ```bash
    fn init
  ```
  func.yaml 파일이 아래와 같은 내용으로 생성된다.
  ```yaml
    version: 0.0.1
    runtime: go
    entrypoint: ./func
  ```
* fn 실행
  ```sh
    sudo fn run
  ```
  실행내용
  ![FN 실행](./fn_run.png)
---

Jupyter NotebookLogout Fn_instroduction_install Last Checkpoint: Yesterday at 3:56 PM (unsaved changes) Python 3
Python 3 Unexpected error while saving file: notebook/Fn_tutorial/chapter1 install/Fn_instroduction_install.ipynb [Errno 2] No such file or directory: '/home/ubuntu/notebook/Fn_tutorial/chapter1 install/Fn_instroduction_install.ipynb'Trusted
File
Edit
View
Insert
Cell
Kernel
Widgets
Help

Introduction To Fn
어떤 밋업 갔다가 알게 되 opensource라서 한번 돌려 보고 싶어서 여기에 쓰는 것임
1. install jupyter kernal for golang
Jupyter에 golang kernel 설치 먼저 함.
origin project url: https://github.com/fnproject/tutorials/blob/master/Introduction/README.md

go get -u github.com/gopherdata/gophernotes

mkdir -p ~/.local/share/jupyter/kernels/gophernotes

cp $GOPATH/src/github.com/gopherdata/gophernotes/kernel/* ~/.local/share/jupyter/kernels/gophernotes
참조 GOPATH가 글로벌로 설정되어있지 않다면 GOPATH 위치를 알아야 하는데 다음과 같은 명령어를 치면 된다.

go env |grep GOPATH
jupyter 재시작해서 커널이 제대로 읽어오는 지 확인 필요!!
만약 go get -u github.com./gopherdata/gophernotes를 할 때 'libzmq was installed under /usr/local/lib, /usr/local/lib/pkgconfig/libzmq.pc exists. PKG_CONFIG_PATH sets to /usr/local/lib/pkgconfig' 처럼 나올 때,

https://stackoverflow.com/questions/24250838/problems-installing-gozmq-with-go-get 참고함

아래 사이트 가서 libzmq를 받아 컴파일 했고, compile에 필요한 build-essential, libtool, autoconf, automake, and uuid-dev을 설치함 http://zeromq.org/docs:source-git

sudo apt install automake
sudo apt install autoconf
sudo apt install libtool
"Kernel error" in a running notebook
커널 에러가 나서 문제가 먼가 했는데, gophernotes가 심볼릭 링크를 못 찾아서 그런 것! 그리고 이런 문제가 많았던지 go kernel 설치하는 github에 README.md 마지막 줄에 아래와 같이 있음.

Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/notebook/base/handlers.py", line 458, in wrapper
    result = yield gen.maybe_future(method(self, *args, **kwargs))
  File "/usr/local/lib/python2.7/site-packages/tornado/gen.py", line 1008, in run
    value = future.result()
  ...
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1335, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
Stop jupyter, if it's already running.

Add a symlink to /go/bin/gophernotes from your path to the gophernotes executable. If you followed the instructions above, this will be:

sudo ln -s $HOME/go/bin/gophernotes /go/bin/gophernotes
Restart jupyter, and you should now be up and running.

2. Docker Install
In [1]:

!curl -fsSL https://get.docker.com/ | sudo sh
# Executing docker install script, commit: 1d31602
+ sh -c apt-get update -qq >/dev/null
+ sh -c apt-get install -y -qq apt-transport-https ca-certificates curl >/dev/null
+ sh -c curl -fsSL "https://download.docker.com/linux/ubuntu/gpg" | apt-key add -qq - >/dev/null
+ sh -c echo "deb [arch=amd64] https://download.docker.com/linux/ubuntu xenial edge" > /etc/apt/sources.list.d/docker.list
+ [ ubuntu = debian ]
+ sh -c apt-get update -qq >/dev/null
+ sh -c apt-get install -y -qq --no-install-recommends docker-ce >/dev/null
+ sh -c docker version
Client:
 Version:	17.12.0-ce
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	c97c6d6
 Built:	Wed Dec 27 20:11:19 2017
 OS/Arch:	linux/amd64

Server:
 Engine:
  Version:	17.12.0-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	c97c6d6
  Built:	Wed Dec 27 20:09:53 2017
  OS/Arch:	linux/amd64
  Experimental:	false
If you would like to use Docker as a non-root user, you should now consider
adding your user to the "docker" group with something like:

  sudo usermod -aG docker your-user

Remember that you will have to log out and back in for this to take effect!

WARNING: Adding a user to the "docker" group will grant the ability to run
         containers which can be used to obtain root privileges on the
         docker host.
         Refer to https://docs.docker.com/engine/security/security/#docker-daemon-attack-surface
         for more information.
3. Fn Download and Install
In [4]:

!curl -LSs https://raw.githubusercontent.com/fnproject/cli/master/install | sh
fn version 0.4.34

        ______
       / ____/___
      / /_  / __ \
     / __/ / / / /
    /_/   /_/ /_/`

4. Fn start
$ sudo fn start
FN 실행


---
## 5. Fn run
### 간단히 go를 통해 예제 돌려 보자!

* go 파일 만들기
  ```bash
    mkdir hello
    cd hello
  ```
* 아래 텍스트를 func.go 생성 후 삽입.
  ```go
    package main
    import {
      "fmt""
    }
    func main(){
      fmt.Println("Hello World!")
    }
  ```
* yaml 생성
  ```bash
    fn init
  ```
  func.yaml 파일이 아래와 같은 내용으로 생성된다.
  ```yaml
    version: 0.0.1
    runtime: go
    entrypoint: ./func
  ```
* fn 실행
  ```sh
    sudo fn run
  ```
  실행내용
  ![FN 실행](./fn_run.png)
---
---
## 5. Fn run
### 간단히 go를 통해 예제 돌려 보자!
​
* go 파일 만들기
  ```bash
    mkdir hello
    cd hello
  ```
* 아래 텍스트를 func.go 생성 후 삽입.
  ```go
    package main
    import {
      "fmt""
    }
    func main(){
      fmt.Println("Hello World!")
    }
  ```
* yaml 생성
  ```bash
    fn init
  ```
  func.yaml 파일이 아래와 같은 내용으로 생성된다.
  ```yaml
    version: 0.0.1
    runtime: go
    entrypoint: ./func
  ```
* fn 실행
  ```sh
    sudo fn run
  ```
  실행내용
  ![FN 실행](./fn_run.png)
---

## 6. fn 생성관련 해석

fn run의 출력이 익숙해지기 전에 Docker를 사용했다면 Dockerfile을 사용하여 docker 빌드를 실행할 때 나타나는 출력과 유사합니다. 물론 이것이 정확히 일어난 일입니다! 이 Fn과 같은 함수를 실행하면 함수에 대한 Dockerfile을 동적으로 생성하고 컨테이너를 작성한 다음이를 실행합니다.(구글번역기 해석)

[원문]
If you have used Docker before the output of fn run should look familiar--it looks like the output you see when running docker build with a Dockerfile. Of course this is exactly what's happening! When you run a function like this Fn is dynamically generating a Dockerfile for your function, building a container, and then running it.

> __NOTE__: Fn 실제로 두 개의 이미지를 사용하고 있습니다. 첫 번째 언어 컴파일러는 바이너리 생성에 사용됩니다. 두 번째 이미지는 생성 된 바이너리와 필요한 언어 런타임 구성 요소 만 패키지합니다. 이 전략을 사용하면 최종 기능 이미지 크기를 가능한 작게 유지할 수 있습니다. 
더 작은 도커 이미지는 자연스럽게 리포지토리 (repository)로부터 push나 pull을 더 빠르므로 전체적인 성능을 향상시킵니다. 이 기술에 대한 자세한 내용은 [작은 이미지 생성을위한 다단계 도커 빌드](https://medium.com/travis-on-docker/multi-stage-docker-builds-for-creating-tiny-go-images-e0e1867efe5a).

> [원문]__NOTE__: Fn is actually using two images.  The first contains the language compiler and is used to generate a binary.  The second image packages only the generated binary and any necessary language runtime components. Using this strategy, the final function image size can be kept as small as possible.
Smaller Docker images are naturally faster to push and pull from a repository which improves overall performance.  For more details on this technique see [Multi-Stage Docker Builds for Creating Tiny Go Images](https://medium.com/travis-on-docker/multi-stage-docker-builds-for-creating-tiny-go-images-e0e1867efe5a).

`fn run`은 로컬 연산입니다. 함수를 빌드하고 패키지화합니다.
로컬 컴퓨터에있는 컨테이너 이미지로 Fn은 다음과 같습니다.
Docker에 내장되어 있으면`docker` cli를 사용하여 지역을 볼 수 있습니다
방금 생성 한 컨테이너 이미지.

[원문] `fn run` is a local operation.  It builds and packages your function
into a container image which resides on your local machine.  As Fn is
built on Docker you can use the `docker` cli to see the local
container image you just generated.


아래 명령어를 입력하면 여러개의 Docker image를 볼 수 있다. 원문처럼은 안나오고(내가 잘못했느지) image가 생성된 것은 볼 수 있다.
사실 ubuntu(oralce cloud)를 사용하다 보니 `fn`든 `docker`든 `sudo`를 함께 사용해야된다.
>`sudo docker images`

```sh
hello                               0.0.1               d64b4a1a15b9        2 minutes ago      6.98MB
```

## 7.  First Function 배포하기
로컬이나 cloud server에 돌릴 수 있다. 만들었던 function을 pulishing하거나 다른 사용자 또는 시스템에서 접근 가능 위해, function 배포가 필요하다. 
> 참고로 `fn start`된 상태에서만 deploy가 가능한 것 같다. 계속 안 되다가 fn server를 실행하니 `sudo fn call myapp /hello`가 되었다.

아래와 같은 명령어를 사용하면 쉽게 배포 가능하다.
 `fn deploy --app myapp --local`


아마 아래와 같은 출력을 볼 수 있을 것이다:

```sh
sudo fn deploy --app myapp --local
Deploying hello to app: myapp at path: /hello
Bumped to version 0.0.5
Building image hello:0.0.5 .....
```

`--app myapp`을 상세히 하므로써 Function이 그룹핑된다. 아래 내용들은 암시적으로 "myapp"으로 function을 사용할 것이다.
`--local` 은 local server로 배포한다는 것을 말하지만, functino image를 registry에 psuh하지 않는다. 필요하다면 원격 Fn server에도 배표가능하다.

Docker build한 것을 보면 `Updating route /hello using image hello:0.0.5`이 출력될 것이다.

이렇게하면 이미지 "hello : 0.0.5"에 패키지 된 기능이 Fn 서버에 의해 "/ hello"경로에 바인딩되었음을 알 수 있다. 아래 경로를 사용하는 방법을 살펴 보겠습니다.

> __NOTE__ : folder 이름에 'hello'을 포함하고 있는 것은 생성된 function container의 이름으로 사용되고, 그 container에 바인딩된 route가 된다.


## 8. Calling Your Deployed Function

배포한 function을 실행하는 방법은 두 개 있다. 첫번째는 `fn` cli를 통한 것이 있다.
>`sudo fn call myapp /hello`

결과는 아래와 같이 나온다.

```sh
ubuntu@fn-test:~/notebook/Fn_tutorial/chapter1_install_run/hello: sudo fn call myapp /hello
Hello from Fn!
```
Local에서 function을 실행할 때 이런 명령어는 바뀌지 않는다. 하지만 "myapp /hello"를 불렀을 때, fn server는 "myapp"를 찾고, "/hello"로 바인딩된 function을 찾을 것이다.

다른 방법은 HTTP를 통해 function을 부르는 것이다. Fn server는 배포한 서버를 "http://localhost:8080/r/myapp/hello" 로 접근할 수 있을 것이다.
curl을 사용해서 해보면(local)
>`curl http://localhost:8080/r/myapp/hello`
>`curl http://server_id:8080/r/myapp/hello`

원격 서버에 돌렸다면 아래와 같이 브라우저에 실행해도 같은 결과를 볼 수 있을 것이다.
>`http://server_id:8080/r/myapp/hello`

결과는 아래와 같이 나올 것이다.
```sh
Hello from Fn!
```
